source/fe/mapping_q_cache.cc

```diff
@@ -249,17 +259,454 @@ MappingQCache<dim, spacedim>::initialize(
+          else
+            result = fe_values_all.get()->get_quadrature_points();
+
+          // for non-active or artificial cells we are done here and return
+          // the absolute positions, since the provided vector cannot contain
+          // any useful information for these cells
+          if (is_active_non_artificial_cell == false)
+            return result;
+        }
+      else
+        {
+          result.resize(
+            Utilities::pow<unsigned int>(this->get_degree() + 1, dim));
+        }
+
+      // Step 2c) read global vector and adjust points accordingly
+      if ((is_fe_q || is_fe_dgq) && fe.degree == this->get_degree())
+        {
+          // case 1: FE_Q or FE_DGQ with same degree as this class has; this
+          // is the simple case since no interpolation is needed
+          std::vector<types::global_dof_index> dof_indices(
+            fe.n_dofs_per_cell());
+          cell_dofs->get_dof_indices(dof_indices);
+
+          for (unsigned int i = 0; i < dof_indices.size(); ++i)
+            {
+              const auto id = fe.system_to_component_index(i);
+
+              if (is_fe_q)
+                {
+                  // case 1a: FE_Q
+                  if (vector_describes_relative_displacement)
+                    result[id.second][id.first] +=
+                      vector_ghosted(dof_indices[i]);
+                  else
+                    result[id.second][id.first] =
+                      vector_ghosted(dof_indices[i]);
+                }
+              else
+                {
+                  // case 1b: FE_DGQ
+                  if (vector_describes_relative_displacement)
+                    result[lexicographic_to_hierarchic_numbering[id.second]]
+                          [id.first] += vector_ghosted(dof_indices[i]);
+                  else
+                    result[lexicographic_to_hierarchic_numbering[id.second]]
+                          [id.first] = vector_ghosted(dof_indices[i]);
+                }
+            }
+        }
+      else
+        {
+          // case 2: general case; interpolation is needed
+          // note: the following code could be optimized for tensor-product
+          // elements via application of sum factorization as is done on
+          // MatrixFree/FEEvaluation
+          auto &fe_values = fe_values_all.get();
+
+          std::vector<Vector<typename VectorType::value_type>> values(
+            fe_values->n_quadrature_points,
+            Vector<typename VectorType::value_type>(spacedim));
+
+          fe_values->get_function_values(vector_ghosted, values);
+
+          for (unsigned int q = 0; q < fe_values->n_quadrature_points; ++q)
+            for (unsigned int c = 0; c < spacedim; ++c)
+              if (vector_describes_relative_displacement)
+                result[q][c] += values[q][c];
+              else
+                result[q][c] = values[q][c];
+        }
+
+      return result;
+    });
```

Comment:
One think I wonder for this function: Now that we have interpolated the values from all active cells, we have a mismatch in the geometry between the active cells and the level cells. We would then typically go on and call the function with the `MGLevelObject`. I wonder if we should try to fix things here: Since we know the geometry by the position of the fine cells from the given vector, can't we use #11804 to perform an arbitrary interpolation to get the positions in the level vectors? My idea is that the levels will be only around 1/7 of the cells on the active level in 3D or around 1/3 in 2D, so maybe one could take the risk to choose a brute-force evaluation. This would allow us to simply skip the other function, and free the user from having to set up the MG data structures, which might be unexpected if you don't know about it.

And if we do a proper job with optimizing #11804, we would not even notice that we do some very heavy lifting here.

Reply:
Implementing this wouldn't be hard (and the code would be definitely easier). But I have concerns regarding:

https://github.com/dealii/dealii/blob/f03ed9ec8651af89c4106058391707ad2d0769a0/source/base/mpi_remote_point_evaluation.cc#L90-L91

and all the `GridTools::find_all_active_cells_around_point()` calls. 

I guess the best approach is to split up this PR so that we can merge at least one function while waiting for #11804.

Reply:
Yes, I had been thinking about this was well. There is also another subtle challenge in this function with the multigrid context that I tacitly ignored before: What if sufficiently many MPI ranks request the value on the same point (or points owned by a single MPI rank)? I think we should aim for eventually solving these issues, but it might be something too challenging for the short term, in which case we might go with the proposed variant for the 9.3 release. So I agree on the general plan.

Reply:
> What if sufficiently many MPI ranks request the value on the same point (or points owned by a single MPI rank)?

What kind of local refinement do you have perform to trigger such a behavior?

Reply:
A uniform mesh suffices - think of a cube where all processes want the mapping information for level 0. Whoever happens to own the unknown in the lower left corner needs to send this information to all ranks. I don't think we have a mechanism (yet) to mediate such requests to the requester via multiple ranks like in a MPI broadcast operations.

Reply:
> think of a cube where all processes want the mapping information for level 0.

I am a bit confused. In out use case it is the opposite around. Coarser levels want information from finer level (actually the finest level); and those should be partitioned well.

Reply:
As far as I remember we set up the coarse level cells as ghosts on all MPI processes, also in case those processes do not own any of the cells on that level and even if none of cells on the first level are own on that process. This implies every single rank requests the positions of the single cell on level 0. Even though the points on that single cell will be well-distributed, whoever process owns that cell will have the burden to send the position to every single rank (unless there is some smartness in #11804 that I don't recall) That would be an overwhelming number with 100k MPI ranks and I see lots of eager messages here. So I see the problem on the other end of the communication that you are describing.

Reply:
> As far as I remember we set up the coarse level cells as ghosts on all MPI processes, also in case those processes do not own any of the cells on that level and even if none of cells on the subsequent level are owned on that process.

If that is the case, yes (I should remember what the distributed triangulations are doing).

But as far as I can see it, we don't need to update the mapping of those cells, do we? Essentially they are "artificial"; but we do not have the notion of a different kind of artificial cells. `MappingQCache::initialize()` could make the distinction and query the position of points only for cells that are actually relevant. But this is all for the future...

*I will try to fix the compilation error on the Windows system in a minute.*

Reply:
> But as far as I can see it, we don't need to update the mapping of those cells, do we? Essentially they are "artificial"; but we do not have the notion of a different kind of artificial cells.

I wonder if we can't simply make them `artificial` for the case where none of the immediate descendants are locally owned? That should avoid this problem in the natural way. Let me try to have a look.

Reply:
> As far as I remember we set up the coarse level cells as ghosts on all MPI processes, also in case those processes do not own any of the cells on that level and even if none of cells on the subsequent level are owned on that process.

It turned out I was wrong - we only ghost those cells where we have an immediate child (as I guessed would be the right thing to do) at least for the parallel distributed triangulation, see https://github.com/dealii/dealii/blob/57e91c245854cc49033bf3e4e13ec8759cffb15c/source/distributed/tria.cc#L2523-L2541
So this seems less critical than I thought for a possible new function with this scheme (we would only want to fill owned or ghosted cells, not the artificial ones).